packages/angular_devkit/build_angular/src/angular-cli-files/utilities/find-specs.ts

```diff
@@ -0,0 +1,31 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import { Path, getSystemPath, join } from '@angular-devkit/core';
+import * as glob from 'glob';
+
+export function findSpecs(input: string, workspaceRoot: Path, sourceRoot: string): string[] {
+
+  input = input.replace(/\\/g, '/'); // normalize input
+
+  // remove source root to support relative paths from root
+  // such paths are easy to get when running script via IDEs
+  if (sourceRoot && input.startsWith(sourceRoot + '/')) {
+    input = input.substr(sourceRoot.length + 1); // +1 to include slash
+  }
+
+  if (input.endsWith('.ts') && !input.endsWith('.spec.ts')) {
+    input = input.substr(0, input.length - 2) + 'spec.ts';
+  } else if (!input.endsWith('.spec.ts') && input.indexOf('*') === -1) {
+    input += '**/*.spec.ts';
+  }
+  const files = glob.sync(input, {
+    cwd: getSystemPath(join(workspaceRoot, sourceRoot)),
```

Comment:
WDYT of resolving the main test file instead, and use it's dir as `cwd`? That way you don't need to assume it's at the sourceRoot.

Reply:
True, I can go with main file directory. I still need to know sourceRoot path though to support input glob relative to workspace root.

Will update

Reply:
Now I remember why I did not use the value of `webpackConfig.entry`, it's type is `string | string[] | Entry | EntryFunc` - I don't know if its safe to assume that I'll get a string. Not sure if you know, but the last one in the list may also return a promise.

Somehow this line that you commented on seems simpler than supporting all variants of entry.

Reply:
Done, now using `dirname(mainFilePath)`